src/modules/launcher/PowerLauncher/ViewModel/MainViewModel.cs

```diff
@@ -440,54 +441,56 @@ private void QueryResults()
                 var currentCancellationToken = _updateSource.Token;
                 _updateToken = currentCancellationToken;
 
-                ProgressBarVisibility = Visibility.Hidden;
                 var query = QueryBuilder.Build(QueryText.Trim(), PluginManager.NonGlobalPlugins);
                 if (query != null)
                 {
-                    // handle the exclusiveness of plugin using action keyword
-                    RemoveOldQueryResults(query);
-
                     _lastQuery = query;
-                    var plugins = PluginManager.ValidPluginsForQuery(query);
-
                     Task.Run(() =>
                     {
-                        // so looping will stop once it was cancelled
-                        var parallelOptions = new ParallelOptions { CancellationToken = currentCancellationToken };
+                        Thread.Sleep(20);
+                        RemoveOldQueryResults(query);
+                        var plugins = PluginManager.ValidPluginsForQuery(query);
+
                         try
                         {
-                            Parallel.ForEach(plugins, parallelOptions, plugin =>
+                            currentCancellationToken.ThrowIfCancellationRequested();
+                            foreach(PluginPair plugin in plugins)
                             {
-                                if (!plugin.Metadata.Disabled)
+                                if (!plugin.Metadata.Disabled && !currentCancellationToken.IsCancellationRequested)
                                 {
                                     var results = PluginManager.QueryForPlugin(plugin, query);
-                                    if (Application.Current.Dispatcher.CheckAccess())
+                                    currentCancellationToken.ThrowIfCancellationRequested();
+                                    lock (_addResultsLock)
                                     {
                                         UpdateResultView(results, plugin.Metadata, query);
```

Comment:
@somil55.  I think we might want to expose the cancelation token to the UpdateResultView, and/or consider having the plugin clamp the number of results.    

The UI is no longer blocked (which is awesome!!).  But if we pass a large list of results into UpdateResultView the `_addResultLock` will still block further queries from completing. 

An easy way to reproduce this.  
1. In the Command window type `C:\Repos\PowerToys\> *.cs`.  -> Notice the result will take a long time to return, but the textbox is now unblocked.
2. Type a new Query 'Microsoft Teams'.  The results will take a really long time. 

Expected results: 
1. Results should return in a reasonable amount of time. 
2. New queries, should cancel older queries. 

With regards to 1.  If we clamp the results.  We can make sure the user get a response in a reasonable amount of time.  A future improvement could be to warn the user the results are being truncated and allow them a way of executing the full search (potentially with a progress bar of some sort).   I think this could be something we improve in a future PR. 

With regards to 2.  Passing in a cancellation token and checking against it in any of the loops executed from within `NewResults` should allow us to cancel the existing query.

Reply:
Added #5261 and #5262 to track these issues.